plugins/woocommerce-admin/client/products/product-page.tsx

```diff
@@ -42,8 +42,22 @@ const ProductMVPFeedbackModalContainer = lazy( () =>
 export default function ProductPage() {
 	const { productId: productIdSearchParam } = useParams();
 
+	/**
+	 * Only register blocks and unregister them when the product page is being rendered or unmounted.
+	 * Note: Dependency array should stay empty.
+	 */
 	useEffect( () => {
 		document.body.classList.add( 'is-product-editor' );
+
+		const unregisterBlocks = initBlocks();
+
+		return () => {
+			document.body.classList.remove( 'is-product-editor' );
+			unregisterBlocks();
+		};
+	}, [] );
```

Comment:
This actually/should fix the React 18 issue that we had looked into fixing: https://github.com/woocommerce/woocommerce/issues/48933

Reason was that we were unregistering blocks when we were changing the route to an existing product for the first time, because it was relying on `productIdSearchParam` see line 109 ( new )